### ISSUE TITLE
gulp.dest() returns before finishing (cb does work)

### ISSUE THREAD

**Issue by hannwong** [gulpjs/gulp#1254]
### `gulp.dest() returns before files are actually copied to destination

Turns out it is not `cb` that doesn't work, but `gulp.dest()` that doesn't complete at the time that its stream is returned to another task.

In my exploration below, I found that listening on `gulp.dest().on('end')` solves this problem.

Incidentally, `del` also returns before finishing. However, it does end properly on a `Promise.then()`, unlike `gulp.dest()` which does not.

### This is outdated

I have 3 tasks: 'clean', 'gather-assets', 'deploy-assets'. And they must be run in that order. Their dependencies are specified accordingly.

The 'clean' task takes in cb, and ends with calling `cb()`.

Same for the 'gather-assets' task. Before calling `cb()`, it calls `gulp.src().pipe(gulp.dest())` 4 times, copying files from 4 different locations.

Task 'deploy-assets' is a single `gulp.src().pipe(gulp.dest())`.

Even task `gather-assets' fails, with error `ENOENT, chmod`. I'm guessing that 'clean' tasks wiped out the files before 'gather-assets' task can do a `chmod`.

Now, if I call 'clean' task manually first, then 'gather-assets' task runs fine. I run 'gather-assets' task right away after that, and I see the above `chmod` error.

**Note:** The above error happens half the time. I counted an average of 5 times out of 10 tries.

Using Gulp 3.9.

I tried `run-sequence`. Different errors, but same non-deterministic (race-condition) outcome.

Note to self: Read more about promises.

**Comment by TrySound** [comment]
@hannwong  Show code.

**Comment by TrySound** [comment]
I think you use the last version of `del` module, which has different api. For now it returns promise and you can just
```
gulp.task('clean', function () {
  return del('path');
});
```

**Comment by hannwong** [comment]
The codes:

    gulp.task('clean', function(cb) {
        del('app');
        del('src/scss');
        del('src/js');
        del(['src/assets/**', '!src/assets', '!src/assets/custom', '!src/assets/custom/**']);
        cb();
    });

    gulp.task('gather-assets', ['clean'], function(cb) {
        // Copy jQuery.
        gulp.src(config.vendor_path + '/app/assets/jquery/**/*',
                {base: config.Con_path + '/app/assets'})
            .pipe(gulp.dest('./src/assets'));

        // Copy Font Awesome.
        gulp.src(config.vendor_path + '/app/assets/font-awesome/**/*',
                 {base: config.Con_path + '/app/assets'})
            .pipe(gulp.dest('src/assets'));

        // Copy js files.
        gulp.src(config.Con_path + '/src/js/**/*', {base: config.Con_path + '/src'})
            .pipe(gulp.dest('src'));

        // Copy scss files.
        gulp.src(config.Con_path + '/src/scss/**/*', {base: config.Con_path + '/src'})
            .pipe(gulp.dest('src'));

        cb();
    });

I have omitted the 3rd step. Even the 2nd step, involving 2 tasks, breaks.

**Comment by TrySound** [comment]
@hannwong del is async function. streams are async too. You end task before del and stream ends.

**Comment by hannwong** [comment]
@TrySound If `del()` is an async function, how do I do a 'clean' task that must be completed before other tasks begin?

You saved a lot of my time by confirming that `del()` is an async function. Thanks!

**Comment by TrySound** [comment]
@hannwong Keep calm and read documentation!
```js
gulp.task('clean', function(cb) {
    return Promise.all([
        del('app'),
        del('src/scss'),
        del('src/js'),
        del(['src/assets/**', '!src/assets', '!src/assets/custom', '!src/assets/custom/**'])
    ]);
});

gulp.task('js', function () {
    // returning a stream inform gulp when it ends.
    return gulp.src(config.Con_path + '/src/js/**/*', {base: config.Con_path + '/src'})
        .pipe(gulp.dest('src'));
});

gulp.task('css', function () {
    return gulp.src(config.Con_path + '/src/scss/**/*', {base: config.Con_path + '/src'})
        .pipe(gulp.dest('src'));
});
```

**Comment by hannwong** [comment]
@TrySound Thanks a lot for your help!

Just to confirm, I don't need to explicitly wrap task 'gather-assets' into a single Promise, do I?

Codes that work perfectly...

```javascript
gulp.task('clean', function() {
    return Promise.all([
        del('app'),
        del('src/scss'),
        del('src/js'),
        del(['src/assets/**', '!src/assets', '!src/assets/tpg', '!src/assets/tpg/**'])
    ]);
});

gulp.task('gather-assets', ['clean'], function(cb) {
    // Copy jQuery.
    gulp.src(config.vendor_path + '/app/assets/jquery/**/*',
            {base: config.vendor_path + '/app/assets'})
    .pipe(gulp.dest('./src/assets'));

    // Copy Font Awesome.
    gulp.src(config.vendor_path + '/app/assets/font-awesome/**/*',
             {base: config.vendor_path + '/app/assets'})
    .pipe(gulp.dest('src/assets'));

    // Copy js files.
    gulp.src(config.vendor_path + '/src/js/**/*', {base: config.vendor_path + '/src'})
    .pipe(gulp.dest('src'));

    // Copy scss files.
    gulp.src(config.vendor_path + '/src/scss/**/*', {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));

    cb();
});

gulp.task('deploy-assets', ['gather-assets'], function() {
    gulp.src('./src/assets/**/*', {base: './src/assets'})
    .pipe(gulp.dest('./app/assets'));
});
```

**Comment by TrySound** [comment]
@hannwong Split gether-assets with small tasks. It's like small modules. Then you can have more control. For now deploy-assets will start before gather-assets ends.

**Comment by hannwong** [comment]
@TrySound I have split task 'gather-assets' into smaller tasks.

But I seem to have come upon another unrelated problem: `gulp.src()` is not finding all the files in `src/assets`.

```javascript
gulp.task('gather-jquery', ['clean'], function(cb) {
    // Copy jQuery.
    gulp.src(config.Con_path + '/app/assets/jquery/**/*',
             {base: config.Con_path + '/app/assets'})
        .pipe(gulp.dest('./src/assets'));
    cb();
});

gulp.task('gather-fontawesome', ['clean'], function(cb) {
    // Copy Font Awesome.
    gulp.src(config.vendor_path + '/app/assets/font-awesome/**/*',
             {base: config.vendor_path + '/app/assets'})
        .pipe(gulp.dest('src/assets')),
    cb();
});

gulp.task('gather-js', ['clean'], function(cb) {
    // Copy js files.
    gulp.src(config.vendor_path + '/src/js/**/*', {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));
    cb();
});

gulp.task('gather-scss', ['clean'], function(cb) {
    // Copy scss files.
    gulp.src(config.vendor_path + '/src/scss/**/*', {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));
    cb();
});

gulp.task('deploy-assets', ['gather-jquery', 'gather-fontawesome', 'gather-js', 'gather-scss'], function(cb) {
    gulp.src('src/assets/**/*', {base: 'src'})
        .pipe(gulp.dest('app'));
    cb();
});
```

I have 3 folders inside folder `src/assets`, but only 1 of them is copied over to `app/assets`.

**Comment by TrySound** [comment]
@hannwong Do not call cb, return streams. Maybe you have wrong sources,

**Comment by TrySound** [comment]
@hannwong clean task calls everytime you use it in dependencies. And why copy a few times? Why not copy once where they should be?

**Comment by hannwong** [comment]
@TrySound How do I return stream when I have 4-5 separate `gulp.src().pipe()` calls?

I have combined the copying into a single step, no more intermediate steps.

```javascript
gulp.task('clean', function() {
    return Promise.all([
        del('app'),
        del('src/scss'),
        del('src/js'),
        del(['src/assets/**', '!src/assets', '!src/assets/tpg', '!src/assets/tpg/**'])
    ]);
});

gulp.task('deploy-assets', ['clean'], function() {
    // Copy jQuery.
    gulp.src(config.vendor_path + '/app/assets/jquery/**/*',
             {base: config.vendor_path + '/app/assets'})
        .pipe(gulp.dest('app/assets'));

    // Copy Font Awesome.
    gulp.src(config.vendor_path + '/app/assets/font-awesome/**/*',
             {base: config.vendor_path + '/app/assets'})
        .pipe(gulp.dest('app/assets'));

    // Copy our own assets.
    gulp.src('src/assets/ours/**/*',
             {base: 'src/assets'});
        .pipe(gulp.dest('app/assets'));

    // Copy js files.
    gulp.src(config.vendor_path + '/src/js/**/*',
             {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));

    // Copy scss files.
    gulp.src(config.vendor_path + '/src/scss/**/*',
                   {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));

    cb();
});

gulp.task('build', ['deploy-assets'], function() {
    runSequence(['css', 'js', 'js-materialize', 'html']);
});
```

**Comment by TrySound** [comment]
@hannwong 

```js
gulp.task('gather-jquery', ['clean'], function() {
    // Copy jQuery.
    return gulp.src(config.Con_path + '/app/assets/jquery/**/*',
             {base: config.Con_path + '/app/assets'})
        .pipe(gulp.dest('./src/assets'));
});

gulp.task('gather-fontawesome', ['clean'], function() {
    // Copy Font Awesome.
    return gulp.src(config.vendor_path + '/app/assets/font-awesome/**/*',
             {base: config.vendor_path + '/app/assets'})
        .pipe(gulp.dest('src/assets')),
});

gulp.task('gather-js', ['clean'], function() {
    // Copy js files.
    return gulp.src(config.vendor_path + '/src/js/**/*', {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));
});

gulp.task('gather-scss', ['clean'], function() {
    // Copy scss files.
    return gulp.src(config.vendor_path + '/src/scss/**/*', {base: config.vendor_path + '/src'})
        .pipe(gulp.dest('src'));
});

gulp.task('deploy-assets', ['clean', 'gather-jquery', 'gather-fontawesome', 'gather-js', 'gather-scss'], function() {
    return gulp.src('src/assets/**/*', {base: 'src'})
        .pipe(gulp.dest('app'));
});
```

**Comment by callumacrae** [comment]
or this https://github.com/gulpjs/gulp/blob/master/docs/recipes/using-multiple-sources-in-one-task.md

**Comment by hannwong** [comment]
@TrySound It appears that `gulp.src().pipe(gulp.dest())` returns success on a `Promise` even though the file copying has not yet completed. I wrapped all 5 of them in my task 'deploy-assets' into a `Promise.all()`.

The only way I can fix this is to block right after `Promise.all()` and poll the file copy destinations until the files are found to be copied over.

Is there something I can do to fix the lack of proper "success" feedback from `gulp.src().pipe(gulp.dest())`? Or am I still missing something?

**Comment by hannwong** [comment]
@callumacrae The destinations are different. But that is a very good recipe! Thanks for the heads up.

**Comment by TrySound** [comment]
@hannwong Stop to invent new. I already said what do you need. And read more documentation.

**Comment by hannwong** [comment]
@TrySound Thanks for your advice! :-)

**Comment by hannwong** [comment]
@contra TrySound has helped me with this issue.

I'm unsubscribing from it now. I can't close it. Just letting you know I don't need this issue anymore.

Thanks!
